### ISSUE TITLE
Use before declaration false positive when combined with binding exports

### ISSUE THREAD

**Issue by Kovensky** [Microsoft/TypeScript#14679]
**TypeScript Version:**  2.2.1

**Code**

```ts
import methodFactory from './http-client'

export { httpDelete as delete } // block scoped httpDelete used before declaration
const httpDelete = methodFactory('DELETE')
```

**Expected behavior:**

No error. The export declaration makes no use of the variable; what should happen is that `httpDelete`'s TDZ should extend to the `delete` export, but that happens regardless of the relative position of the `export` declaration.

**Actual behavior:**

Two errors in the same line about the variable being used before assignment.

This error is correct in the case of `export default` and the typescript-specific `export =`, but not in any other use of `export`.

**Comment by kitsonk** [comment]
It is use before declaration.  The error is valid.  The export declaration *does* make use of the variable as there is an expected assignment.  You are making assumptions about how the code is loaded in a loader.  While in practice currently, a module loader may not run into a TDZ issue, in practice, the block scoped variable is being used before it is assigned, which is against the language spec.

**Comment by aluanhaddad** [comment]
Indeed, note that there is no error when using `var`, and that this will work correctly when using a target module format and loader that properly support live bindings.

**Comment by Kovensky** [comment]
@kitsonk no, that is literally how it is supposed to work according to spec. [No evaluation should happen](https://tc39.github.io/ecma262/#sec-exports-runtime-semantics-evaluation): `1. Return NormalCompletion(empty).`.

If any evaluation _does_ happen, it is a bug with your module loader.

**Comment by Kovensky** [comment]
@aluanhaddad it will work correctly, which is why it should not be a compile error.

Even the code generated by the playground [is (more-or-less) correct](http://www.typescriptlang.org/play/index.html#src=export%20%7B%20foo%20%7D%0D%0Aconst%20foo%20%3D%20'bar'%0D%0A), but it will raise a compilation error anyway.

**Comment by RyanCavanaugh** [comment]
Simple enough fix to not issue this error when the parent is an export declaration

**Comment by aluanhaddad** [comment]
@Kovensky that's a great point, and I imagine it's one of the reasons that the register format uses `var` even one targeting es2015. However when TypeScript transpiles to CommonJS, under the es2015 target, it neither uses `var` nor generates a getter. It just generates an assignment to the `exports` object. As you say this behavior is not correct. Several issues have been logged, #12522 and #13245 among others, regarding the current behavior.

Thank you for correcting me and for linking to the spec.

**Comment by gcnew** [comment]
Does this apply to classes as well?

With the current nightly (`typescript@2.3.0-dev.20170403`) I get an error for exporting classes before their declaration.

```ts
export { XYZ }
//       ~~~
// XYZ.ts(1,10): error TS2449: Class 'XYZ' used before its declaration.

class XYZ {
}
```

**Comment by RyanCavanaugh** [comment]
@gcnew that should be valid too. Almost certainly the same root cause.

Should be an easy PR if anyone's interested

**Comment by gcnew** [comment]
I haven't tried it, but it seems that @Kovensky's PR (https://github.com/Microsoft/TypeScript/pull/14700) takes care for classes as well. I just saw that he has commented that class exports were miraculously working - https://github.com/Microsoft/TypeScript/pull/14700#discussion_r106568485.

**Comment by gcnew** [comment]
The issue still persists if you compile with `--strict`, works fine without :).
